packages/grid/veilid/server/veilid_streamer.py

```diff
@@ -0,0 +1,480 @@
+           ```
+           async def handle_receive_stream(message: bytes) -> bytes:
+               # Do something with the message once the entire stream is received.
+               return b'some response to the sender of the stream.'
+           ```
+
+        3. Add the following to your connection's update_callback function to relay
+           updates to the VeilidStreamer properly:
+           ```
+           def update_callback(update: veilid.VeilidUpdate) -> None:
+               if vs.is_stream_update(update):
+                   vs.receive_stream(connection, update, handle_receive_stream)
+               ...other callback code...
+           ```
+
+        4. Use the `stream` method to send an app_call with a message of any size.
+           ```
+           response = await vs.stream(router, vld_key, message)
+           ```
+
+    Special case:
+        If the message is small enough to fit in a single chunk, we can send it as a
+        single STREAM_SINGLE request. This avoids the additional overhead while still
+        allowing large replies containing multiple chunks.
+
+        stream_single_struct = Struct("!8s16s")  # 24 bytes
+            [RequestType.STREAM_SINGLE (8 bytes string)] +
+            [Call ID (16 bytes random UUID string)] +
+
+        Therefore, the maximum size of the message that can be sent in a STREAM_SINGLE
+        request is 32768 - 24 = 32744 bytes.
+            [stream_single_struct (24 bytes)] +
+            [Actual Message (32744 bytes)]
+            = 32768 bytes
+
+        Data flow for single chunk message:
+            Sender side:
+                1. Send STREAM_SINGLE request -> Get OK
+                2. Await reply from the receiver
+                3. Return the reply once received
+            Receiver side:
+                1. Get STREAM_SINGLE request -> Send OK
+                2. Pass the message to the callback function and get the reply
+                3. Stream the reply back to the sender
+
+        Usage:
+            This is automatically handled by the VeilidStreamer class. You don't need to
+            do anything special for this. Just use the `stream` method as usual. If the
+            message is small enough to fit in a single chunk, it will be sent as a
+            STREAM_SINGLE request automatically.
+    """
+
+    _instance = None
+    buffers: dict[CallId, Buffer]
+
+    def __new__(cls) -> "VeilidStreamer":
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+            cls._instance.buffers = {}
+        return cls._instance
+
+    def __init__(self) -> None:
+        # Structs for serializing and deserializing metadata as bytes of fixed length
+        # https://docs.python.org/3/library/struct.html#format-characters
+        BYTE_ORDER = "!"  # big-endian is recommended for networks as per IETF RFC 1700
+        REQUEST_TYPE_PREFIX = f"{RequestType.SIZE}s"
+        CALL_ID = "16s"
```

Comment:
MAYBE: On reading the code, since veilid also has a call_id attribute for each app call and we use it in functions like _send_response. To make it easy for new users navigating the code.

We could re-name  our `call_id` to `stream_id` ?

Reply:
This was super needed 🙌